Branch-SDK/src/io/branch/referral/ServerRequest.java

```diff
@@ -388,41 +388,57 @@ private void updateGAdsParams() {
+                if (userDataObj != null) {
+                    userDataObj.put(Defines.Jsonkey.LimitedAdTracking.getKey(), LATVal);
+                    if (!TextUtils.isEmpty(gaid)) {
                         userDataObj.put(Defines.Jsonkey.AAID.getKey(), gaid);
-                        userDataObj.put(Defines.Jsonkey.LimitedAdTracking.getKey(), LATVal);
                         userDataObj.remove(Defines.Jsonkey.UnidentifiedDevice.getKey());
+                        // for future use
+                        keyToBeUsedInAdIdsObject = getKeyToBeUsedInAdIdsObject(userDataObj);
+                    } else if (payloadContainsNoDeviceIdentifiers(userDataObj)) {
+                        userDataObj.put(Defines.Jsonkey.UnidentifiedDevice.getKey(), true);
                     }
-                } else {
-                    params_.put(Defines.Jsonkey.GoogleAdvertisingID.getKey(), gaid);
-                    params_.put(Defines.Jsonkey.LATVal.getKey(), LATVal);
                 }
-            } catch (JSONException e) {
-                e.printStackTrace();
-            }
-        } else { // Add unidentified_device when neither GAID nor AndroidID are present
-            if (version == BRANCH_API_VERSION.V2 || version == BRANCH_API_VERSION.V1_CPID) {
-                try {
-                    if (version == BRANCH_API_VERSION.V2 || version == BRANCH_API_VERSION.V1_CPID ||
-                            version == BRANCH_API_VERSION.V1_LATD) {
-                        JSONObject userDataObj = params_.optJSONObject(Defines.Jsonkey.UserData.getKey());
-                        if (userDataObj != null) {
-                            if (!userDataObj.has(Defines.Jsonkey.AndroidID.getKey())) {
-                                userDataObj.put(Defines.Jsonkey.UnidentifiedDevice.getKey(), true);
-                            }
-                        }
-                    }
-                } catch (JSONException ignore) {
-                
+            } else {
+                if (!TextUtils.isEmpty(gaid)) {
+                    params_.put(Defines.Jsonkey.GoogleAdvertisingID.getKey(), gaid);
+                    // for future use
+                    keyToBeUsedInAdIdsObject = getKeyToBeUsedInAdIdsObject(params_);
+                } else if (payloadContainsNoDeviceIdentifiers(params_)) {
+                    params_.put(Defines.Jsonkey.UnidentifiedDevice.getKey(), true);
                 }
+                params_.put(Defines.Jsonkey.LATVal.getKey(), LATVal);
             }
+
+            if (keyToBeUsedInAdIdsObject == null) return;
+
+            JSONObject advertisingIdsObject = new JSONObject().put(keyToBeUsedInAdIdsObject, gaid);
+            params_.put(Defines.Jsonkey.AdvertisingIDs.getKey(), advertisingIdsObject);
+        } catch (JSONException e) {
+            e.printStackTrace();
         }
     }
+
+    private boolean payloadContainsNoDeviceIdentifiers(JSONObject payload) {
```

Comment:
Can we reverse this logic to say `payloadContainsDeviceIdentifiers()` and do the "not" check just once?